File: 2708-find-the-string-with-lcp/find-the-string-with-lcp.java
class Solution {
    public String findTheString(int[][] lcp) {
        int n = lcp.length;

        // Basic validation: diagonal must be correct and matrix must be symmetric
        for (int i = 0; i < n; i++) {
            if (lcp[i][i] != n - i) return "";
            for (int j = 0; j < n; j++) {
                if (lcp[i][j] != lcp[j][i]) return "";
            }
        }

        char[] word = new char[n];
        for (int i = 0; i < n; i++) word[i] = '#';

        char ch = 'a';

        // Greedily assign smallest possible characters
        for (int i = 0; i < n; i++) {
            if (word[i] != '#') continue;

            if (ch > 'z') return ""; // more than 26 groups needed

            for (int j = i; j < n; j++) {
                if (lcp[i][j] > 0) {
                    word[j] = ch;
                }
            }
            ch++;
        }

        // Verify by recomputing LCP matrix from constructed string
        int[][] dp = new int[n + 1][n + 1];

        for (int i = n - 1; i >= 0; i--) {
            for (int j = n - 1; j >= 0; j--) {
                if (word[i] == word[j]) {
                    dp[i][j] = 1 + dp[i + 1][j + 1];
                } else {
                    dp[i][j] = 0;
                }

                if (dp[i][j] != lcp[i][j]) return "";
            }
        }

        return new String(word);
    }
}

File: 2708-find-the-string-with-lcp/find-the-string-with-lcp.cpp
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    string findTheString(vector<vector<int>>& lcp) {
        int n = lcp.size();
        string word(n, ' ');
        char currentChar = 'a';
        
        // Step 1: Greedily construct the lexicographically smallest string
        for (int i = 0; i < n; ++i) {
            if (word[i] == ' ') {
                if (currentChar > 'z') {
                    return ""; // More than 26 distinct characters needed
                }
                for (int j = i; j < n; ++j) {
                    if (lcp[i][j] > 0) {
                        word[j] = currentChar;
                    }
                }
                currentChar++;
            }
        }
        
        // Step 2: Validate the constructed string against the given LCP matrix
        // We traverse backwards to use the already validated lcp[i+1][j+1] state natively
        for (int i = n - 1; i >= 0; --i) {
            for (int j = n - 1; j >= 0; --j) {
                int expectedLcp = 0;
                
                if (word[i] == word[j]) {
                    if (i + 1 < n && j + 1 < n) {
                        expectedLcp = lcp[i + 1][j + 1] + 1;
                    } else {
                        expectedLcp = 1; // Out of bounds means suffix length is exactly 1
                    }
                }
                
                if (lcp[i][j] != expectedLcp) {
                    return ""; // Found a contradiction, matrix is invalid
                }
            }
        }
        
        return word;
    }
};